File: maps/src/main/java/it/localhostsoftware/maps/MapFragment.java
package it.localhostsoftware.maps;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import com.google.android.gms.common.GoogleApiAvailability;
import com.huawei.hms.api.HuaweiApiAvailability;

import it.localhostsoftware.maps.google.GoogleMap;
import it.localhostsoftware.maps.huawei.HuaweiMap;

public class MapFragment extends Fragment {
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_map, container, false);
        Fragment fragment;
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(requireContext()) == com.google.android.gms.common.ConnectionResult.SUCCESS)
            fragment = com.google.android.gms.maps.SupportMapFragment.newInstance();
        else if (HuaweiApiAvailability.getInstance().isHuaweiMobileServicesAvailable(requireContext()) == com.huawei.hms.api.ConnectionResult.SUCCESS)
            fragment = com.huawei.hms.maps.SupportMapFragment.newInstance();
        else throw new IllegalStateException();
        getChildFragmentManager().beginTransaction().replace(R.id.fragmentContainerView, fragment).commitAllowingStateLoss();
        return v;
    }

    public void getMapAsync(OnMapReadyCallback callback) {
        getChildFragmentManager().registerFragmentLifecycleCallbacks(new FragmentManager.FragmentLifecycleCallbacks() {
            @Override
            public void onFragmentViewCreated(@NonNull FragmentManager fragmentManager, @NonNull Fragment f, @NonNull View view, @Nullable Bundle savedInstanceState) {
                Fragment fragment = getChildFragmentManager().findFragmentById(R.id.fragmentContainerView);
                if (fragment instanceof com.google.android.gms.maps.SupportMapFragment) {
                    fragmentManager.unregisterFragmentLifecycleCallbacks(this);
                    ((com.google.android.gms.maps.SupportMapFragment) fragment).getMapAsync(googleMap -> callback.onMapReady(new GoogleMap(googleMap)));
                } else if (fragment instanceof com.huawei.hms.maps.SupportMapFragment) {
                    fragmentManager.unregisterFragmentLifecycleCallbacks(this);
                    ((com.huawei.hms.maps.SupportMapFragment) fragment).getMapAsync(googleMap -> callback.onMapReady(new HuaweiMap(googleMap)));
                }
            }
        }, false);
    }
}

File: maps/src/main/java/it/localhostsoftware/maps/MapFragment.kt
package it.localhostsoftware.maps

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.BundleCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import it.localhostsoftware.maps.google.GoogleMap
import it.localhostsoftware.maps.google.GoogleMapOptions
import it.localhostsoftware.maps.huawei.HuaweiMap
import it.localhostsoftware.maps.huawei.HuaweiMapOptions

class MapFragment : Fragment() {
    companion object {
        private const val MAP_OPTIONS = "mapOptions"
        fun newInstance(mapOptions: MapOptions<*, *, *>?) = MapFragment().apply {
            arguments = Bundle().apply {
                if (mapOptions is GoogleMapOptions) putParcelable(MAP_OPTIONS, mapOptions.mo)
                else if (mapOptions is HuaweiMapOptions) putParcelable(MAP_OPTIONS, mapOptions.mo)
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View = inflater.inflate(R.layout.fragment_map, container, false).also {
        childFragmentManager.beginTransaction().replace(
            R.id.fragmentContainerView, when (requireContext().getMobileServices()) {
                MobileServices.GOOGLE -> {
                    if (arguments?.containsKey(MAP_OPTIONS) == true) com.google.android.gms.maps.SupportMapFragment.newInstance(BundleCompat.getParcelable(requireArguments(), MAP_OPTIONS, com.google.android.gms.maps.GoogleMapOptions::class.java))
                    else com.google.android.gms.maps.SupportMapFragment.newInstance()
                }

                MobileServices.HUAWEI -> {
                    if (arguments?.containsKey(MAP_OPTIONS) == true) com.huawei.hms.maps.SupportMapFragment.newInstance(BundleCompat.getParcelable(requireArguments(), MAP_OPTIONS, com.huawei.hms.maps.HuaweiMapOptions::class.java))
                    else com.huawei.hms.maps.SupportMapFragment.newInstance()
                }

                else -> throw IllegalStateException()
            }
        ).commitAllowingStateLoss()
    }

    fun getMapAsync(block: (GeoMap<*, *, *, *, *, *, *, *, *, *, *, *, *, *, *, *, *, *, *, *, *>) -> Unit) {
        childFragmentManager.registerFragmentLifecycleCallbacks(object : FragmentManager.FragmentLifecycleCallbacks() {
            override fun onFragmentViewCreated(fragmentManager: FragmentManager, f: Fragment, view: View, savedInstanceState: Bundle?) {
                childFragmentManager.findFragmentById(R.id.fragmentContainerView).let { mapFragment ->
                    if (mapFragment is com.google.android.gms.maps.SupportMapFragment) {
                        fragmentManager.unregisterFragmentLifecycleCallbacks(this)
                        mapFragment.getMapAsync { block(GoogleMap(it)) }
                    } else if (mapFragment is com.huawei.hms.maps.SupportMapFragment) {
                        fragmentManager.unregisterFragmentLifecycleCallbacks(this)
                        mapFragment.getMapAsync { block(HuaweiMap(it)) }
                    }
                }
            }
        }, false)
    }
}